src/ParallelAlgorithms/LinearSolver/Schwarz/ElementCenteredSubdomainData.hpp

```diff
@@ -0,0 +1,214 @@
+  data *= scalar;
+  return data;
+}
+
+template <size_t Dim, typename TagsList>
+decltype(auto) operator*(ElementCenteredSubdomainData<Dim, TagsList> data,
+                         const double scalar) noexcept {
+  data *= scalar;
+  return data;
+}
+
+template <size_t Dim, typename TagsList>
+decltype(auto) operator/(ElementCenteredSubdomainData<Dim, TagsList> data,
+                         const double scalar) noexcept {
+  data /= scalar;
+  return data;
+}
+
+template <size_t Dim, typename TagsList>
+std::ostream& operator<<(std::ostream& os,
+                         const ElementCenteredSubdomainData<Dim, TagsList>&
+                             subdomain_data) noexcept {
+  os << "Element data:\n"
+     << subdomain_data.element_data << "\nOverlap data:\n"
+     << subdomain_data.overlap_data;
+  return os;
+}
+
+template <size_t Dim, typename LhsTagsList, typename RhsTagsList>
+bool operator==(
+    const ElementCenteredSubdomainData<Dim, LhsTagsList>& lhs,
+    const ElementCenteredSubdomainData<Dim, RhsTagsList>& rhs) noexcept {
+  return lhs.element_data == rhs.element_data and
+         lhs.overlap_data == rhs.overlap_data;
+}
+
+template <size_t Dim, typename LhsTagsList, typename RhsTagsList>
+bool operator!=(
+    const ElementCenteredSubdomainData<Dim, LhsTagsList>& lhs,
+    const ElementCenteredSubdomainData<Dim, RhsTagsList>& rhs) noexcept {
+  return not(lhs == rhs);
+}
+
+}  // namespace LinearSolver::Schwarz
+
+namespace LinearSolver::InnerProductImpls {
+
+template <size_t Dim, typename LhsTagsList, typename RhsTagsList>
+struct InnerProductImpl<
+    Schwarz::ElementCenteredSubdomainData<Dim, LhsTagsList>,
+    Schwarz::ElementCenteredSubdomainData<Dim, RhsTagsList>> {
+  static double apply(
+      const Schwarz::ElementCenteredSubdomainData<Dim, LhsTagsList>& lhs,
+      const Schwarz::ElementCenteredSubdomainData<Dim, RhsTagsList>&
+          rhs) noexcept {
+    double result = inner_product(lhs.element_data, rhs.element_data);
+    for (const auto& id_and_lhs_overlap_data : lhs.overlap_data) {
+      result +=
+          inner_product(id_and_lhs_overlap_data.second,
+                        rhs.overlap_data.at(id_and_lhs_overlap_data.first));
+    }
+    return result;
+  }
+};
+
+}  // namespace LinearSolver::InnerProductImpls
+
+namespace MakeWithValueImpls {
+
+template <size_t Dim, typename TagsListOut, typename TagsListIn>
+struct MakeWithValueImpl<
+    LinearSolver::Schwarz::ElementCenteredSubdomainData<Dim, TagsListOut>,
+    LinearSolver::Schwarz::ElementCenteredSubdomainData<Dim, TagsListIn>> {
+  using SubdomainDataIn =
+      LinearSolver::Schwarz::ElementCenteredSubdomainData<Dim, TagsListIn>;
+  using SubdomainDataOut =
+      LinearSolver::Schwarz::ElementCenteredSubdomainData<Dim, TagsListOut>;
+  static SPECTRE_ALWAYS_INLINE SubdomainDataOut
+  apply(const SubdomainDataIn& input, const double value) noexcept {
+    SubdomainDataOut output{input.element_data.number_of_grid_points()};
```

Comment:
can you default construct this instead of passing in the number of grid points? I'm concerned this allocates twice when it could just allocate once on the next line.